File: backend/node_version/src/controllers/userController.ts
// controllers/userController.ts
import { Request, Response } from 'express';
import { addUser, deleteUsers, findUserByEmail, assignRoleToUser, getUsers as fetchUsers} from '../services/userService';

export const createUser = async (req: Request, res: Response) => {
  const { email, name, role } = req.body;

  try {
    const user = await addUser(email, name, role);
    res.json(user);
  } catch (error: any) {
    res.status(400).json({ message: error.message });
  }
};

export const removeUsers = async (req: Request, res: Response) => {
  const { userIds } = req.body; // Expecting an array of user IDs

  // Validate input
  if (!userIds || !Array.isArray(userIds)) {
    res.status(400).json({ message: 'userIds is required and should be an array.' });
  }

  try {
    await deleteUsers(userIds);
    res.json({ message: 'Users deleted successfully' });
  } catch (error: any) {
    res.status(400).json({ message: error.message });
  }
};


export const getUserByEmail = async (req: Request, res: Response) => {
  const { email } = req.params;

  try {
    await findUserByEmail(email);
    res.json({ message: 'User found successfully' });
  } catch (error: any) {
    res.status(500).json({ message: error.message });
  }
};

export const changeUserRole = async (req: Request, res: Response) => {
  const { userId } = req.params;
  const { role } = req.body;

  try {
    await assignRoleToUser(userId, role);
    res.json({ message: 'Role updated successfully' });
  } catch (error: any) {
    res.status(500).json({ message: error.message });
  }
};

// Fetch all users
export const getUsers = async (req: Request, res: Response) => {
  try {
    const users = await fetchUsers();
    res.json(users);
  } catch (error: any) {
    res.status(500).json({ message: error.message });
  }
};

File: backend/scala-seed-project/src/controllers/UserController.scala
// src/controllers/UserController.scala

import cats.effect.Sync
import org.http4s.{Request, Response}
import org.http4s.dsl.Http4sDsl
import org.http4s.HttpRoutes
import io.circe.generic.auto._
import io.circe.syntax._
import services.UserService

case class CreateUserRequest(email: String, name: String, role: String)
case class UserIdsRequest(userIds: List[Int])
case class ChangeUserRoleRequest(role: String)

object UserController {
  def routes[F[_]: Sync](userService: UserService[F]): HttpRoutes[F] = {
    val dsl = new Http4sDsl[F]{}
    import dsl._

    HttpRoutes.of[F] {
      // Create user handler
      case req @ POST -> Root / "users" =>
        for {
          createUserRequest <- req.as[CreateUserRequest]
          user <- userService.addUser(createUserRequest.email, createUserRequest.name, createUserRequest.role)
          resp <- Ok(user.asJson)
        } yield resp

      // Remove users handler
      case req @ DELETE -> Root / "users" =>
        for {
          userIdsRequest <- req.as[UserIdsRequest]
          _ <- userService.deleteUsers(userIdsRequest.userIds)
          resp <- Ok("Users deleted successfully")
        } yield resp

      // Get user by email handler
      case req @ GET -> Root / "users" / email =>
        for {
          user <- userService.findUserByEmail(email)
          resp <- Ok(user.asJson)
        } yield resp

      // Change user role handler
      case req @ PUT -> Root / "users" / IntVar(userId) =>
        for {
          changeUserRoleRequest <- req.as[ChangeUserRoleRequest]
          _ <- userService.assignRoleToUser(userId, changeUserRoleRequest.role)
          resp <- Ok("Role updated successfully")
        } yield resp

      // Fetch all users handler
      case GET -> Root / "users" =>
        for {
          users <- userService.getUsers()
          resp <- Ok(users.asJson)
        } yield resp
    }
  }
}
